src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        return getElementByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> current = getElementByIndex(index);
+        T result = current.value;
+        current.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        T result;
+        if (index == 0 && size == 1) {
+            result = head.value;
+            head = null;
+            tail = null;
+            node = null;
+            size--;
+            return result;
+        }
+        Node<T> current = getElementByIndex(index);
+        if (current.prev == null) {
+            result = current.value;
+            head = current.next;
+            current.next.prev = null;
+            current.next = null;
+            size--;
+            return result;
+        }
+        if (current.next == null) {
+            result = current.value;
+            tail = current.prev;
+            current.prev.next = null;
+            current.prev = null;
+            size--;
+            return result;
+        }
+        result = current.value;
+        current.prev.next = current.next;
+        current.next.prev = current.prev;
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 1 && head.value == object) {
+            head = null;
+            tail = null;
+            node = null;
+            size--;
+            return true;
+        }
+        Node<T> current = head;
+        while (current != null) {
+            if (current.value == null && current.value == object || current.value.equals(object)) {
+                if (current.prev == null) {
+                    head = current.next;
+                    current.next.prev = null;
+                    current.next = null;
+                    size--;
+                    return true;
+                }
+                if (current.next == null) {
+                    tail = current.prev;
+                    current.prev.next = null;
+                    current.prev = null;
+                    size--;
+                    return true;
+                }
+                current.prev.next = current.next;
+                current.next.prev = current.prev;
+                size--;
+                return true;
```

Comment:
let's create a private method `unlink(Node<T> node)` and when you find node - pass it to this method
move unlinking logic to that method